### PR TITLE
[SVS-24] Technical - refactoring the controller and bind/connect commands

### PR DESCRIPTION
**Pull request by eyalantzMS**
See description in the original PR that I closed in order to produce this PR which would be easier to review: https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/pull/13

### CODE REVIEW

File: src/Integration/Binding/BindCommand.cs [hunk 1]
```diff
@@ -1,159 +0,0 @@
...
-using SonarLint.VisualStudio.Integration.TeamExplorer;
-using SonarLint.VisualStudio.Integration.WPF;
-using SonarLint.VisualStudio.Progress.Controller;
-using System;
-using System.Diagnostics;
-using System.Linq;
-
-namespace SonarLint.VisualStudio.Integration.Binding
-{
-    internal class BindCommand : HostedCommandBase, IBindingWorkflowExecutor
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindCommand.cs]: The BindCommand/ConnectCommand were renamed to BindController/ConnectionController.

Comment (eyalantzMS) [inline src/Integration/Binding/BindCommand.cs]: General comment about tests - test coverage is at least as previously. In some cases what we had as an integration tests were replaced by unit tests and interfaces to mock and simplify the new integration tests.

File: src/Integration/ArchetectureOverview.dgml [hunk 2]
```diff
@@ -0,0 +1,360 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (eyalantzMS) [inline src/Integration/ArchetectureOverview.dgml]: This is the dgml i used to create the image on the previous version of this PR

File: src/Integration/Binding/BindingController.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+using System.Diagnostics;
+using System.Linq;
+using System.Windows.Input;
+
+namespace SonarLint.VisualStudio.Integration.Binding
+{
+    /// <summary>
+    /// A dedicated controller for the <see cref="BindCommand"/>
+    /// </summary>
+    internal class BindingController : HostedCommandControllerBase, IBindingWorkflowExecutor
```
Comment (eyalantzMS) [inline src/Integration/Binding/BindingController.cs]: There's binding controller and connection controller, each of them expose commands which are aggregated at the top level section controller and used for UI databinding. Some of the command could be hidden when they are not bound to the UI (i.e. Don't show again).
The lifecycle of those controller is tied to the section controller, where as the state spans across all the controllers in a VS session and as such defined and managed by other classes that were introduced part of this refactoring (they are the main reason for the refactoring really).

File: src/Integration/Connection/ConnectionWorkflow.cs [hunk 4]
```diff
@@ -26,17 +25,20 @@ namespace SonarLint.VisualStudio.Integration.Connection
     /// </summary>
     internal class ConnectionWorkflow
```
Comment (eyalantzMS) [inline src/Integration/Connection/ConnectionWorkflow.cs]: Another goal was to break the dependency of the workflows on the controllers whilst still consuming the needed services which are now exposed by the host.

File: src/Integration/TeamExplorer/SectionController.cs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    {
+        public const string SectionId = "25AB05EF-8132-453E-A990-55587C0C5CD3";
+        public const int Priority = 300;
+
+        internal const int CommandNotHandled = (int)OleConstants.OLECMDERR_E_UNKNOWNGROUP;
+
+        private readonly IWebBrowser webBrowser;
+
+        [ImportingConstructor]
+        public SectionController([Import] IHost host, IWebBrowser webBrowser)
```
Comment (mjcheetham) [inline src/Integration/TeamExplorer/SectionController.cs]: This constructor needs test for null arguments (i.e., a MEF failure); `this.Host` and this.webBrowser` are assumed to never be null in this class.
Also, add the `[Import]` attribute to both ctor arguments (or remove them).

File: src/Integration/TeamExplorer/SectionController.cs [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+            this.Host = host;
+            this.webBrowser = webBrowser;
+        }
+
+        internal /*for testing purposes*/ List<IOleCommandTarget> CommandTargets
+        {
+            get;
+        } = new List<IOleCommandTarget>();
+
+        internal /*for test purposes*/ IHost Host
```
Comment (mjcheetham) [inline src/Integration/TeamExplorer/SectionController.cs]: This property is only used externally from this class twice, in the same test helper method `SectionControllerTests.ReInitialize`. All the SectionController tests use the `CreateTestSubject` helper which `new`s-up a test host. Could the test host not be a field in the TestClass and interact directly with it in the ReInitialize helper, therefore removing the need for the `internal` access modifier here.

File: src/Integration.UnitTests/TeamExplorer/SectionControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            public void AssertQueryStatusCalled(int expectedNumberOfTimes)
+            {
+                Assert.AreEqual(expectedNumberOfTimes, this.queryStatusNumberOfCalls, "IOleCommandTarget.QueryStatus is called unexpected number of times");
+            }
+            #endregion
+        }
+
+        private SectionController CreateTestSubject(IWebBrowser webBrowser = null)
+        {
+            var host = new ConfigurableHost(this.serviceProvider, Dispatcher.CurrentDispatcher);
```
Comment (mjcheetham) [inline src/Integration.UnitTests/TeamExplorer/SectionControllerTests.cs]: Related to my other comment about `internal IHost` on `SectionController`, we could have this `ConfigurableHost` created at test init and then referenced here & in `ReInitialize`; removing the need for the `internal` access modifier.

File: src/TestInfrastructure/Framework/ConfigurableConnectSection.cs [hunk 8]
```diff
@@ -5,14 +5,53 @@
...
 
+using SonarLint.VisualStudio.Integration.Progress;
 using SonarLint.VisualStudio.Integration.TeamExplorer;
-using System.Windows;
+using SonarLint.VisualStudio.Integration.WPF;
+using System.Windows.Input;
 
 namespace SonarLint.VisualStudio.Integration.UnitTests
 {
     internal class ConfigurableConnectSection : IConnectSection
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableConnectSection.cs]: Did we not want to also rename this class (and interface) to match the concrete `SectionController`? I.e., `ConfigurableSectionController` and `ISectionController`.

File: src/Integration.UnitTests/State/StateManagerTests.cs [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+        [TestMethod]
+        public void StateManager_SetProjectsUIThread()
+        {
+            // Setup
+            var section = ConfigurableConnectSection.CreateDefault();
+            ConfigurableUserNotification notifications = (ConfigurableUserNotification)section.UserNotifications;
+            ConfigurableHost host = new ConfigurableHost();
+            StateManager testSubject = this.CreateTestSubject(host);
+            host.VisualStateManager = testSubject;
+            section.ViewModel.State = testSubject.ManagedState;
```
Comment (mjcheetham) [inline src/Integration.UnitTests/State/StateManagerTests.cs]: We do this (`section.ViewModel.State = testSubject.ManagedState`) in every test setup, move it in to the create helper.

File: src/Integration/State/StateManager.cs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+            }
+
+            if (state == null)
+            {
+                throw new ArgumentNullException(nameof(state));
+            }
+
+            this.Host = host;
+            this.ManagedState = state;
+            this.ManagedState.PropertyChanged += this.OnStatePropertyChanged;
```
Comment (mjcheetham) [inline src/Integration/State/StateManager.cs]: Why not have an `IsBusyChanged` event rather than a general event which means the subscriber will have to check if it's interested or not? Especially since we're only exposing `IsBusy` at the moment and presumably the specific property being changed will necessitate a different action; we don't need to respond to all changes uniformly?

File: src/Integration/TeamExplorer/SectionController.cs [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        private bool CanToggleShowAllProjects(ServerViewModel server)
+        {
+            return server.Projects.Any(x => x.IsBound);
+        }
+
+        private void ToggleShowAllProjects(ServerViewModel server)
+        {
+            server.ShowAllProjects = !server.ShowAllProjects;
+        }
+        private bool CanExecBrowseToUrl(string url)
```
Comment (mjcheetham) [inline src/Integration/TeamExplorer/SectionController.cs]: Missing newline between methods.

File: src/Integration/Connection/ConnectionController.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            this.IsConnectionInProgress = true;
+
+            ProgressNotificationListener progressListener = new ProgressNotificationListener(this.ServiceProvider, progressEvents);
+            progressListener.MessageFormat = Strings.ConnectingToSonarQubePrefixMessageFormat;
+
+            progressEvents.RunOnFinished(result =>
+            {
+                progressListener.Dispose();
+                this.IsConnectionInProgress = false;
+                this.ShowNuGetWarning(result);
```
Comment (mjcheetham) [inline src/Integration/Connection/ConnectionController.cs]: Agreed, this feels like a better place to invoke the warning, rather than as a progress step on the workflow.

File: src/Integration.UnitTests/Connection/ConnectControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+        }
+
+        #region  Tests
+        [TestMethod]
+        public void ConnectionController_Ctor_ArgumentChecks()
+        {
+            ConnectionController suppressAnalysisIssue;
+            Exceptions.Expect<ArgumentNullException>(() =>
+            {
+                suppressAnalysisIssue = new ConnectionController(null);
```
Comment (mjcheetham) [inline src/Integration.UnitTests/Connection/ConnectControllerTests.cs]: Could also avoid the analysis issue by replacing the code block with an expression (`() => new ConnectionController(null)`). I was confused at first seeing this variable.

File: src/Integration/Binding/BindingController.cs [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            this.projectSystemHelper = projectSystemHelper ?? new ProjectSystemHelper(this.host);
+        }
+
+        public RelayCommand<ProjectViewModel> BindCommand
+        {
+            get;
+        }
+
+        #region Command
+        internal bool IsBindingInProgress
```
Comment (mjcheetham) [inline src/Integration/Binding/BindingController.cs]: /*for testing purposes*/

File: src/Integration/Binding/BindingWorkflow.cs [hunk 15]
```diff
@@ -95,7 +95,7 @@ public List<NuGetPackageInfo> NuGetPackages
 
         public IProgressEvents Run()
         {
-            this.owner.UserNotification.HideNotification(NotificationIds.FailedToBindId);
+            this.host.ActiveSection?.UserNotifications?.HideNotification(NotificationIds.FailedToBindId);
```
Comment (mjcheetham) [inline src/Integration/Binding/BindingWorkflow.cs]: This is redundant since we're already clearing the notification in `BindingController.OnBindingStarted()`

File: src/Integration/IHost.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+using System.Windows.Threading;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    internal interface IHost : IServiceProvider
+    {
+        /// <summary>
+        /// The UI thread dispatcher. Not null.
+        /// </summary>
+        Dispatcher UIDIspatcher { get; }
```
Comment (mjcheetham) [inline src/Integration/IHost.cs]: Typo: the `I` in Dispatcher has been upper-cased.

File: src/Integration/VsSessionHost.cs [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+        public void ClearActiveSection()
+        {
+            if (this.ActiveSection == null) // Can be called multiple times
+            {
+                return;
+            }
+
+            this.ActiveSection.ViewModel.State = null;
+            this.ActiveSection = null;
+            this.VisualStateManager.SyncCommandFromActiveSection();
```
Comment (mjcheetham) [inline src/Integration/VsSessionHost.cs]: Should this call to `SyncCommandFromActiveSection` happen here?
When we navigate away from the section in the Team Explorer, the SectionController is disposed, which in turn calls this method. `SyncCommandFromActiveSection` then actually goes off to create new context menu commands (which aren't accessible until we re-attach an active section).

File: src/Integration/TeamExplorer/SectionController.cs (hunk 6 above)
Comment (eyalantzMS) [inline src/Integration/TeamExplorer/SectionController.cs]: Yep, good catch.

File: src/Integration/VsSessionHost.cs (hunk 17 above)
Comment (eyalantzMS) [inline src/Integration/VsSessionHost.cs]: Should it happen here - yes. We don't want to reference dead instances which will prevent them from being GCed.
Should we created (disabled) commands - no. (fixed that)

File: src/Integration/Binding/BindingWorkflow.cs (hunk 15 above)
Comment (eyalantzMS) [inline src/Integration/Binding/BindingWorkflow.cs]: Definitely.

**Comment by eyalantzMS** [comment]
Addressed all comments.

**Comment by mjcheetham** [comment]
LGTM
